From 79df60ca93f554ed2078ffb855e4f1633cbd2dab Mon Sep 17 00:00:00 2001
From: Arjen Hiemstra <ahiemstra@heimr.nl>
Date: Tue, 30 Sep 2025 11:53:37 +0200
Subject: [PATCH] kcm: Restore runtime check for OSTree backend

It's possible to build discover with the OSTree backend while running on
a system that doesn't actually use OSTree. This would mean there's no
option for selecting how system updates work even though the system
supports choosing that, as the current logic is purely tied to a compile
time check. This restores the runtime check if the OSTree backend was
built. The systemd case is still based on a compile-time check as I
don't know how to check that at runtime.

BUG: 510072
---
 kcm/CMakeLists.txt | 10 ++++++----
 kcm/updates.cpp    |  8 +++++++-
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/kcm/CMakeLists.txt b/kcm/CMakeLists.txt
index 745f8ceb8..a3c895f7c 100644
--- a/kcm/CMakeLists.txt
+++ b/kcm/CMakeLists.txt
@@ -26,8 +26,10 @@ target_link_libraries(kcm_updates PRIVATE
     KF6::KCMUtilsQuick
 )
 
-if(BUILD_RpmOstreeBackend OR BUILD_SystemdSysupdateBackend)
-    target_compile_definitions(kcm_updates PRIVATE UPDATE_NEEDS_REBOOT=1)
-else()
-    target_compile_definitions(kcm_updates PRIVATE UPDATE_NEEDS_REBOOT=0)
+if (BUILD_RpmOstreeBackend)
+    target_compile_definitions(kcm_updates PRIVATE WITH_OSTREE_BACKEND)
+endif()
+
+if (BUILD_SystemdSysupdateBackend)
+    target_compile_definitions(kcm_updates PRIVATE WITH_SYSUPDATE_BACKEND)
 endif()
diff --git a/kcm/updates.cpp b/kcm/updates.cpp
index a269ca3b9..28c4eb68e 100644
--- a/kcm/updates.cpp
+++ b/kcm/updates.cpp
@@ -40,7 +40,13 @@ DiscoverSettings *Updates::discoverSettings() const
 
 bool Updates::mandatoryRebootAfterUpdate() const
 {
-    return UPDATE_NEEDS_REBOOT;
+#if defined WITH_SYSUPDATE_BACKEND
+    return true;
+#elif defined WITH_OSTREE_BACKEND
+    return QFile::exists(QStringLiteral("/run/ostree-booted"));
+#else
+    return false;
+#endif
 }
 
 #include "moc_updates.cpp"
-- 
GitLab

